File: Python/pointer/grumpy_bookstore_owner.py
from typing import List

def maxSatisfied(customers: List[int], grumpy: List[int], X: int) -> int:
    length = len(grumpy)
    left, right, imax, remain, total = 0, 0, 0, 0, 0
    while right < X:
        if grumpy[right] == 0:
            total += customers[right]
        else:
            remain += customers[right]
        right += 1

    imax = max(remain, imax)
    while right < length:
        if grumpy[right] == 0:
            total += customers[right]
        else:
            remain += customers[right]
        if grumpy[left] == 1:
            remain -= customers[left]
        imax = max(imax, remain)
        left += 1
        right += 1

    
    return total + imax

if __name__ == "__main__":
    print(maxSatisfied(customers = [1,0,1,2,1,1,7,5], grumpy = [0,1,0,1,0,1,0,1], X = 3))

File: src/tmp/Java/src/com/dbc/algorithm/Pointer/GrumpyBookstoreOwner.java
package com.dbc.algorithm.Pointer;

public class GrumpyBookstoreOwner {
    public static int maxSatisfied(int[] customers, int[] grumpy, int X) {
        int left = 0, right = 0, total = 0, remain = 0;
        while (right < X) {
            if (grumpy[right] == 0) {
                total += customers[right];
            } else {
                remain += customers[right];
            }
            right++;
        }
        int max = remain;
        while (right < grumpy.length) {
            if (grumpy[right] == 0) {
                total += customers[right];
            } else {
                remain += customers[right];
            }
            if (grumpy[left] == 1) {
                remain -= customers[left];
            }
            max = Math.max(max, remain);
            left++;
            right++;
        }
        return total + max;
    }
}
